controllers/front/ajax.php

```diff
@@ -0,0 +1,60 @@
+<?php
+/**
+ * 2007-2018 PrestaShop
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Academic Free License (AFL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * http://opensource.org/licenses/afl-3.0.php
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   http://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+if (!defined('_PS_VERSION_')) {
+    exit;
+}
+
+/**
+ * Class DateofdeliveryAjaxModuleFrontController
+ */
+class DateofdeliveryAjaxModuleFrontController extends ModuleFrontController
+{
+    /** @var DateOfDelivery $module */
+    public $module;
+
+    /**
+     * @throws PrestaShopException
+     * @throws SmartyException
+     */
+    public function initContent()
+    {
+        parent::initContent();
+        @ob_clean();
+        header('Content-Type: application/json;charset=UTF-8');
+
+        $cart = $this->context->cart;
+        if (!$cart->id_carrier) {
+            die(json_encode(['success' => false]));
```

Comment:
Can you use ajaxRender instead `die`?

Reply:
I honestly never understood why you would do that. So many modules for PrestaShop are broken, and are hooked to ajaxDie when they shouldn't. Now with the transition to PHP 7.2/7.3 in particular you are risking the output buffer from getting clogged with errors by modules and messing up the response. Is that okay nowadays?

If we want modules to influence the output we should come up with new hooks, not this crap.

Reply:
I have made modules for PrestaShop for many year, but have honestly never seen a necessity to hook onto ajaxRender.

I understand that you want one place where you can control the cache. It is not happening in this module atm, we can adjust that, but it is also not cleaning the output buffer before echoing.

Best would be to log the output buffer, then ditch it, since there's a chance it contains errors. If we output it directly, the JSON can't be parsed. But that is a second issue that arises with this function. It is not made for JSON only, whereas that is the only thing we want to output in our module's controller. Again, it's why we shouldn't be using such a generic hook.

Since we can implement this function in the module's controller: how about making this one JSON only, call the hooks (for those who like to use them), clean the output buffer, then output the JSON response?

Reply:
@PierreRambaud What do you think?

Reply:
@PierreRambaud, if we replace `die` with `ajaxRender`, we break the compatibility with old versions of PrestaShop. I'd keep them here.

Reply:
With MVC you should not be hacking ajax into a regular `FrontController` to begin with, but the most used "alternative" is just calling a plain .php file directly, even worse.
These are all just legacy problems, let's just keep it that way and find something better in the future.

Reply:
And besides, does it really matter for this pull request? All modules have some way or another to provide ajax functionality. As long as there is not a solution that covers all situations, we shouldn't be dealing with it, I think. Just my €0,02